lib/rust/parser/src/main.rs

```diff
@@ -0,0 +1,835 @@
+            // Multiple non-operators next to each other.
+            let lhs = token_to_ast(lhs);
+            let rhs = token_to_ast(item);
+            let ast = syntax::Tree::app(lhs, rhs);
+            output.push(ast.into());
+        } else {
+            // Non-operator that follows previously consumed operator.
+            last_token_was_ast = true;
+            last_token_was_opr = false;
+            output.push(item);
+        }
+    }
+    let mut opt_rhs = last_token_was_ast.and_option_from(|| output.pop().map(token_to_ast));
+    while let Some(opr) = operator_stack.pop() {
+        let opt_lhs = output.pop().map(|t| token_to_ast(t));
+        opt_rhs = Some(syntax::Tree::opr_app(opt_lhs, opr.elem, opt_rhs));
+    }
+    if !output.is_empty() {
+        panic!("Internal error. Not all tokens were consumed while constructing the expression.");
+    }
+    syntax::Tree::opr_section_boundary(opt_rhs.unwrap()) // fixme
+}
+
+
+
+impl<'s, 't, T> Debug for source::With<'s, &'t Box<T>>
+where source::With<'s, &'t T>: Debug
+{
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let t: &T = &**self.data;
+        Debug::fmt(&self.with_data(t), f)
+    }
+}
+
+
+
+impl<'s, 't, T> Debug for source::With<'s, &'t Option<T>>
+where source::With<'s, &'t T>: Debug
+{
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        Debug::fmt(&self.data.as_ref().map(|t| self.with_data(t)), f)
+    }
+}
+
+impl<'s, 't, T, E> Debug for source::With<'s, &'t Result<T, E>>
+where
+    source::With<'s, &'t T>: Debug,
+    source::With<'s, &'t E>: Debug,
+{
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        Debug::fmt(&self.data.as_ref().map(|t| self.with_data(t)).map_err(|t| self.with_data(t)), f)
+    }
+}
+
+
+fn token_to_ast(elem: syntax::Item) -> syntax::Tree {
+    match elem {
+        syntax::Item::Token(token) => match token.elem {
+            token::Type::Ident(ident) =>
+                elem.span().with(syntax::tree::Type::from(syntax::tree::Ident(ident))),
+            _ => panic!(),
+        },
+        syntax::Item::Tree(ast) => ast,
+    }
+}
+
+fn matched_segments_into_multi_segment_app<'s>(
+    lexer: &Lexer<'s>,
+    prefix_tokens: Option<Vec<syntax::Item>>,
+    matched_segments: NonEmptyVec<(Token, Vec<syntax::Item>)>,
+) -> syntax::Tree {
+    // FIXME: remove into_vec
+    let segments = matched_segments
+        .into_vec()
+        .into_iter()
+        .map(|segment| {
+            let header = segment.0;
+            let body = (segment.1.len() > 0)
+                .as_some_from(|| resolve_operator_precedence(lexer, segment.1));
+            syntax::tree::MultiSegmentAppSegment { header, body }
+        })
+        .collect_vec();
+    if let Ok(segments) = NonEmptyVec::try_from(segments) {
+        let prefix = prefix_tokens.map(|t| resolve_operator_precedence(lexer, t));
+        syntax::Tree::multi_segment_app(prefix, segments)
+    } else {
+        panic!()
+    }
```

Comment:
This looks like a case for `NonEmptyVec::mapped`.

Reply:
Added to fixme!